provider/provider.go

```diff
@@ -65,18 +69,134 @@ type KadClosestPeersRouter interface {
 
 type SweepingProvider struct {
 	// TODO: complete me
+	done      chan struct{}
+	closeOnce sync.Once
+
 	peerid peer.ID
+	order  bit256.Key
 	router KadClosestPeersRouter
 
 	clock clock.Clock
 
 	maxProvideConnsPerWorker int
 
+	schedule   *trie.Trie[bitstr.Key, time.Duration]
+	scheduleLk sync.Mutex
+
+	avgPrefixLenLk       sync.Mutex
+	avgPrefixLenReady    chan struct{}
+	cachedAvgPrefixLen   int
+	lastAvgPrefixLen     time.Time
+	avgPrefixLenValidity time.Duration
+
 	msgSender      pb.MessageSender
 	getSelfAddrs   func() []ma.Multiaddr
 	addLocalRecord func(mh.Multihash) error
 }
 
+// Close stops the provider and releases all resources.
+func (s *SweepingProvider) Close() {
+	s.closeOnce.Do(func() { close(s.done) })
+}
+
+func (s *SweepingProvider) closed() bool {
+	select {
+	case <-s.done:
+		return true
+	default:
+		return false
+	}
+}
+
+const initialGetClosestPeers = 4
+
+// measureInitialPrefixLen makes a few GetClosestPeers calls to get an estimate
+// of the prefix length to be used in the network.
+//
+// This function blocks until GetClosestPeers succeeds or the provider is
+// closed. No provide operation can happen until this function returns.
+func (s *SweepingProvider) measureInitialPrefixLen() {
+	cplSum := atomic.Int32{}
+	cplSamples := atomic.Int32{}
+	wg := sync.WaitGroup{}
+	wg.Add(initialGetClosestPeers)
+	for range initialGetClosestPeers {
+		go func() {
+			defer wg.Done()
+			bytes := [32]byte{}
+			rand.Read(bytes[:])
+			for {
+				if s.closed() {
```

Comment:
This seems inefficient to select on a channel here. Maybe move this outside the loop and have `Close` wait for any calls to `measureInitialPrefixLen` to complete.

Reply:
The go routine tries `GetClosestPeers()` until it returns some peers. It means that if the node is offline, it blocks until it comes online (or retries forever).

Hence, `Close()` cannot wait for it to complete since it may hang forever. `measureInitialPrefixLen` go routines should check whether the `SweepingProvider` is closed to exit the infinite loop (in the case the node is offline).

Reply:
It may not be the best approach, but it is critical to make this prefix length measurement before providing anything, otherwise the initial provide may result in a full network crawl (taking ~10 minutes).

Anyway, if `GetClosestPeers` errors (e.g because node is offline), the provider won't be able to provide anything, because it needs `GetClosestPeers` for the provide operation. So it should be fine to hang provide opeartions until `GetClosestPeers` succeeds.